### PR TITLE
HPCC-31293 Refactor code to preserve the thread context

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Initial push to run regression suite.

File: system/jlib/jthread.hpp [hunk 1]
```diff
@@ -362,4 +376,26 @@ interface IWorkQueueThread: extends IInterface
...
 extern jlib_decl IWorkQueueThread *createWorkQueueThread(unsigned persisttime=1000*60);
 
+//--- Functions that manage the current thread state =- for context, tracing, spans etc.
+
+class LogMsgJobInfo;
+
+const LogMsgJobInfo & checkDefaultJobInfo(const LogMsgJobInfo & _jobInfo);
+extern jlib_decl void setDefaultJobId(LogMsgJobId id);
+
+extern const IContextLogger * queryActiveContextLogger();
```
Comment (jakesmith) [inline system/jlib/jthread.hpp:386]: wondering for clarity, if this wouldn't be better named queryThreadedContextLogger()

and (see other comment) re. the 2 varieties of "setDefaultJobId" (one in jlog one in jthread) and naming.

**Review by jakesmith — COMMENTED** [review]
@ghalliday - it does look cleaner/clearer. Please see minor comments.

File: system/jlib/jthread.hpp [hunk 2]
```diff
@@ -33,6 +33,23 @@
 // no thread priority handling?
 #endif
 
+// Functions used to reset thread-local context variables, when a threadpool starts
+interface IContextLogger;
+typedef unsigned __int64 LogMsgJobId;
```
Comment (rpastrana) [inline system/jlib/jthread.hpp:38]: Is it fair to refactor to LogMsgJobInfoId ?
It helps disambiguate between the jobInfo's identifier, and the actual jobId

File: system/jlib/jthread.cpp [hunk 3]
```diff
@@ -378,13 +378,7 @@ void Thread::init(const char *_name)
...
 void Thread::getThreadLoggingInfo()
 {
-    ::getThreadLoggingInfo(logctx, traceFlags);
-}
-
-void Thread::setThreadLoggingInfo(const IContextLogger *_logctx, TraceFlags _traceFlags)
-{
-    logctx = _logctx;
-    traceFlags = _traceFlags;
+    ::getThreadLoggingInfo(savedCtx);
```
Comment (rpastrana) [inline system/jlib/jthread.cpp:381]: Does the savedCtx need to be initialized at Thread::init time?

**Review by rpastrana — COMMENTED** [review]
@ghalliday the change looks good, I pointed out a couple of improvements and an important question regarding the initial state of Thread's savedCtx

**Comment by ghalliday** [comment]
I have made some changes, but the more I think about it, the less I think the current scheme works as you want.
As Rodrigo points out, when a thread is initialised it should possibly save the state.

I think there are two different cases for starting a thread
i) A short term thread which will be terminated before the calling process finishes.  (Thread pools are often a common case).  In this case the context should be inherited from the calling thread.
ii) A long term thread.  In this case it should use a global configuration.

Currently threadpools use the context when they are created, threads generally use the default context.  (That means that many places that previously were initialised with the global job id will now have no job id - so this change will be a backwards step).

I think the correct solution should be something like:
The constructor for Thread should have a (non-optional) boolean parameter to indicate whether it should save the current context, or use the global default context.  This means background threads will not be left with pointers to (or ids of) objects that no longer exist.

Should there also be an option on start() to use a context to override the default context.  This allows a threadpool to be globally shared, but for each instance to match the current context.

This is beginning to feel like a change that is outside the scope of 9.4.x, but I also think it is probably the correct approach/preparation for cleanly adding trace ids to the logging.  That may mean that associating trace ids with logging is also a 9.6 feature.
@rpastrana @jakesmith for further discussion, probably including @richardkchapman

**Comment by ghalliday** [comment]
@jakesmith @rpastrana this should now per a pure-refactoring of the code with no functional changes.
Once this is merged I will investigate changing the way the job ids are updated as a completely separate phase.

File: system/jlib/jlog.cpp [hunk 4]
```diff
@@ -3356,47 +3310,9 @@ IRemoteLogAccess *queryRemoteLogAccessor()
...
-    threadTraceFlags = traceFlags;
-    prev = default_thread_logctx;
-    default_thread_logctx = ctx;
-}
-LogContextScope::~LogContextScope()
+void setDefaultJobId(const char *id)
 {
-    default_thread_logctx = prev;
-    threadTraceFlags = prevFlags;
+    setDefaultJobId(theManager->addJobId(id));
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply system/jlib/jlog.cpp]: I tried that change, but then backed it out.  I suspect the correct change would be setDefautJobName and setDefaultJobId, but I will not make that change as part of this PR.

**Review by rpastrana — APPROVED** [review]

File: system/jlib/jthread.cpp [hunk 5]
```diff
@@ -2683,4 +2670,102 @@ void PerfTracer::dostop()
...
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+#include "jlog.hpp"
+
+static LogMsgJobInfo globalDefaultJobInfo(UnknownJob, UnknownUser);
+
+// NOTE - extern thread_local variables are very inefficient - don't be tempted to expose the variables directly
+
+static  TraceFlags defaultTraceFlags = TraceFlags::Standard;
```
Comment (jakesmith) [inline system/jlib/jthread.cpp:2682]: trivial: double space.

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good to me (1 trivial spacing issue)

**Pull request merged**